5단계 - 로또(수동) 리뷰 요청드립니다

tlaqk229 opened the pull request:
안녕하세요!
4단계 미션 리뷰 시 코멘트 주신 부분 반영 및 5단계 미션 구현해보았습니다.
이번 미션 진행하면서 보너스볼 중복 여부 체크 기능 추가해보았고
보너스볼 관련하여 당첨 카운트 로직에 오류를 발견하여 수정했습니다.
수정하면서 보너스볼 일치여부 Enum 클래스를 추가하고 이를 활용하는 방식으로 개선해봤습니다.

그리고 한가지 질문드릴게 있습니다!
4단계 미션 완료 후 피드백 영상 보고 정적 팩토리 메소드에 대해 찾아보고 학습해보았는데요.
찾아본 내용 중에 정적 팩토리 메소드를 사용함으로써 객체 생성을 캡슐화 할 수 있다는 내용이 있었는데요.
(https://lovethefeel.tistory.com/m/130 본문 중 '4. 객체 생성을 캡슐화할 수 있습니다' 부분)
아래 코드와 같이 객체의 생성자는 private으로 선언하고, 정적 팩토리 메소드에 인자로 Write 객체를 넘겨주어
메소드 내에서 필요한 값을 꺼내와 private 생성자를 호출함으로써
WriteResponse 생성자의 인자들을 숨길 수 있다는 내용인데요.
```java
public class WriteResponse {

    private Long id;
    private String title;
    private String content;
    private LocalDateTime created;
    private String createdBy;

    private WriteResponse(final Long id, final String title, final String content, final LocalDateTime created
            , final String createdBy) {
        this.id = id;
        this.title = title;
        this.content = content;
        this.created = created;
        this.createdBy = createdBy;
    }

    public static WriteResponse from(final Write write) {
        return new WriteResponse(write.getId(), write.getTitle(), write.getContent(), write.getCreated(), write.getCreatedBy());
    }
}
```

아래와 같이 생성자를 하나 더 만들고 인자로 Write 객체를 받게 하는 것과
정적 팩토리 메소드에서 Write 객체를 받아 처리하는 것 사이에 차이가 있는지 의문의 들어 질문드립니다!
새로운 생성자를 만들지 않고 정적 팩토리 메소드를 만들어 사용하는 경우에 얻을 수 있는 장점은 어떤 것이 있는지 궁금합니다.
```java
public WriteResponse(final Write write) {
    this(write.getId(), write.getTitle(), write.getContent(), write.getCreated(), write.getCreatedBy());
}
```

감사합니다!

tlaqk229 committed:
fix(MatchCount): 보너스볼 일치관련 로직 오류 수정

- 보너스볼이 일치하지 않아도 되는 경우에 보너스볼이 일치하면
  당첨으로 카운트되지 않는 오류 수정
- 당첨기준(WinCriterion)에 bonusCount가 0인 경우
  당첨여부 판단에 영향주지 않도록 수정

tlaqk229 committed:
fix(LottoNumber): 정적 팩토리 메소드 추가하여 캐싱 적용

- 한정된 범위 내에서 반복적으로 사용되는 값이므로 캐싱되도록 수정

tlaqk229 committed:
fix(MoneyFromat): Formatter final선언 및 인스턴스 생성 불가처리

tlaqk229 committed:
docs(mission5): 5단계 미션 요구사항 & 기능목록 작성

tlaqk229 committed:
docs(mission5): 기능목록 추가

tlaqk229 committed:
feat(view): 수동 로또 관련 입력 기능 추가

tlaqk229 committed:
feat(PurchaseCount): 구입 개수 정보 객체 추가

- 총 구입 개수, 수동 로또 구입 개수 저장
- 구입 개수 관련 유효성 체크 로직 구현

tlaqk229 committed:
fix(ResultView): 구매 개수 출력 수정

- 수동, 자동 구매 개수 구분하여 출력하도록 수정

tlaqk229 committed:
feat(ErrorMessage): 구매 개수 유효성 체크 관련 오류메세지 추가

tlaqk229 committed:
feat(manualLotto): 수동 로또 발급, 출력 구현

tlaqk229 committed:
fix(BonusBallMatch): 보너스볼 체크 방식 수정

- 보너스볼 일치 여부 확인 Enum 추가
- 당첨 등수 확인 시 BonusBallMatch Enum 사용하도록 수정

tlaqk229 committed:
refactor(all): 자바 코드 컨벤션 - code reformat

tlaqk229 committed:
docs(학습); 정적 팩토리 메소드 학습 내용 정리

ohtaeg commented:
> 아래와 같이 생성자를 하나 더 만들고 인자로 Write 객체를 받게 하는 것과
정적 팩토리 메소드에서 Write 객체를 받아 처리하는 것 사이에 차이가 있는지 의문의 들어 질문드립니다!
새로운 생성자를 만들지 않고 정적 팩토리 메소드를 만들어 사용하는 경우에 얻을 수 있는 장점은 어떤 것이 있는지 궁금합니다.

개인적으로 객체 생성시 데이터를 어디선가 getter 등을 통해 가져와서 해줘야하기 때문에 캡슐화는 큰 차이가 없다고 생각합니다 :)


개인적으로 정적 팩토리 메서드는 생성자가 여러개가 있을 경우 효과를 발휘하는 것 같아요 :)
현업에서 일을 하다보면 많은 필드가 담긴 클래스를 다루게 되는데
필드가 많다보니 각 생성자마다 필요한 데이터들이 다르기에 여러 생성자가 생기게 됩니다.
어떤 상황에 따라 어떤 값을 가져야하는지 정적 팩터리 메서드를 통해 의사전달을 할 수 있다고 생각하며

정적 팩토리 메서드를 사용하면 얻는 또다른 장점은 객체 생성을 호출 할 때마다 인스턴스를 새로 생성하지 않아도 되는 점이라고 생각합니다!
위에서 말씀드린 것 처럼 클래스 규모가 커서 생성 비용이 큰 객체의 경우 요청될 때마다 생성하는 것보다 
정적 팩토리 메소드를 사용하면 매번 인스턴스를 새로 생성하지 않고 
기존에 만들어두거나 생성한 인스턴스를 캐싱해서 재활용할 수 있는 장점이 있다고 생각합니다 :)
디자인 패턴의 플라이웨이트 패턴과 유사한 장점이 있지 않을까 합니다!

원하시는 대답이 되었는지 모르겠네요 🙏

ohtaeg commented on src/main/java/lotto/model/domain/Lotto.java:
```diff
@@ -13,6 +13,18 @@ public class Lotto {
     private static final int LOTTO_NUMBER_DO_NOT_EXIST = 0;
     private List<LottoNumber> lotto;
 
+    public Lotto() {
+    }
```
이 기본 생성자가 꼭 필요한 생성자인지 고민해보시면 좋을 것 같아요 :)

tlaqk229 replied:
기본 생성자는 LottoTest, WinLottoTest, LottoServiceImpl의 generateOneLotto()에서 사용하고 있었는데요. LottoTest에서 비효율적으로 반복되는 Lotto 객체 생성 코드를 개선하기 위해 새로운 생성자를 작성하고 대체하니 LottoServiceImpl의 generateOneLotto()에서의 호출만 남게 되었습니다. 메소드를 보며 생각해보니 LottoService보다는 Lotto 도메인 객체 자체의 로직에 해당된다는 생각이 들어 해당 메소드를 Lotto 객체 안에 구현하고 기본 생성자는 제거하였습니다! (메소드를 옮기면서 이름을 generateOneAutoLotto로 수정하였는데요. 3단계 미션 진행 시에는 자동, 수동 구분이 없어 메소드명을 generateOneLotto라고 작성했는데 조금 더 명확하게 자동로또 생성 메소드임을 드러내는 게 좋을 듯하여 수정해보았습니다)

ohtaeg replied:
의견 공유 감사합니다 😄 
말씀해주신대로 기본 생성자를 생성하는 생성 책임들을 가진 객체들을 살펴보았을 때 위와 같은 고민을 해주시면 좋을 것 같아 드린 피드백이였는데 의도를 정확하게 잘 파악해주셨습니다 👍

ohtaeg commented on src/main/java/lotto/model/domain/Lotto.java:
```diff
@@ -13,6 +13,18 @@ public class Lotto {
+        for (String lottoNumber : lottoNumbers) {
+            lotto.add(LottoNumber.getLottoNumberByString(lottoNumber));
+        }
+        checkLottoNumberCount();
```
개인적으로 생성을 하기 전에 전처리하는 validation이 수행되고나서 값을 세팅해주는게 올바른 순서이지 않을까 생각해요 :)
checkLottoNumberCount() 에 파라미터를 받는 식으로 변경하면 어떨까 싶은데 윤지님의 의견이 궁금하네요!

tlaqk229 replied:
저도 처음에는 checkLottoNumberCount()를 번호 추가 전에 호출해서 지정된 개수(6개)를 초과하는지 유효성 체크를 하도록 작성했었는데요. 문득 드는 생각이, 6개를 초과해서도 안되지만, 부족해서도 안된다고 생각이 들었습니다. 그래서 LottoNumber 추가를 완료하고 부족한지 체크하는 유효성 체크를 하나 더 추가할까 하다가, 초과와 부족을 구분 짓는 것이 꼭 필요할까 싶어서 유효성 체크 시점을 생성자 가장 마지막 단계로 옮기고, Lotto에 추가된 ottoNumber의 개수가 정확히 지정한 개수와 일치하는지 체크하는 방식으로 수정해보았습니다. 초과하는 경우와 부족한 경우를 구분하여 체크하도록하고 초과 체크를 LottoNumber 추가 이전에 호출, 부족 체크를 마지막에 하도록 수정하는 것이 좋을까요? 조언 부탁드립니다!

ohtaeg replied:
보통 생성을 해주거나 값을 세팅해줄때는 유효성 체크를 먼저 해주는 편이 좋지 않을까해서 드린 코멘트 였습니다 :D
만약 생성하고 유효성 검증을 실패해버리면 불필요한 인스턴스가 생성되어 메모리를 잡아먹기 때문인데요!
이런 관점에서 checkLottoNumberCount()는 오로지 6개가 아닐 경우만 체크를 해주고 있기에 lottoNumbers의 갯수로 먼저 유효성을 검증하고 값을 세팅해주면 좋지 않을까 했습니다 😄

ohtaeg replied:
보통 생성을 해주거나 값을 세팅해줄때는 유효성 체크를 먼저 해주는 편이 좋지 않을까해서 드린 코멘트 였습니다 :D
만약 생성하고 유효성 검증을 실패해버리면 짧은 순간이지만 불필요한 인스턴스가 생성되어 메모리를 잡아먹기 때문인데요!
이런 관점에서 checkLottoNumberCount()는 오로지 6개가 아닐 경우만 체크를 해주고 있기에 lottoNumbers의 갯수로 먼저 유효성을 검증하고 값을 세팅해주면 좋지 않을까 했습니다 😄 

초과하는 경우와 부족한 경우는 이미 checkLottoNumberCount()를 통해 요구사항을 만족해주고 있기에
따로 체크안해주셔도 될 것 같아요!

tlaqk229 replied:
추가하기 전에 String[] lottoNumbers의 개수로 유효성 체크를 할 수 있다는 걸 생각 못했네요!! 유효성 검증을 먼저 실행하는 것이 불필요한 인스턴스 생성을 막아준다는 점을 짚어주시니 유효성 검증 순서의 중요성이 더 와닿았습니다! 숙지하고 코딩해야겠다는 생각이 드네요! 조언주신대로 인자를 받아서 검증하는 방식으로 수정해봤습니다!

ohtaeg commented on src/main/java/lotto/model/domain/LottoNumber.java:
```diff
@@ -1,20 +1,36 @@
     private int lottoNumber;
 
-    public LottoNumber(int lottoNumber) {
+    static {
+        for (int i = LottoConstants.LOTTO_NUMBER_MIN; i <= LottoConstants.LOTTO_NUMBER_MAX; i++) {
+            lottoNumbers.put(i, new LottoNumber(i));
+        }
+    }
```
캐싱 💯

ohtaeg commented on src/main/java/lotto/model/domain/LottoNumber.java:
```diff
@@ -1,20 +1,36 @@
 
-    public LottoNumber(String lottoNumber) {
+    public static LottoNumber getLottoNumberByInt(int lottoNumber) {
+        if (lottoNumbers.get(lottoNumber) == null) {
+            return new LottoNumber(lottoNumber);
```
캐시된 데이터가 없다면 새로 생성해주지말고 validateNumber() 를 수행해주면 어떨까요 :) ?

tlaqk229 replied:
이 부분 이해가 잘 안돼서 부연설명 부탁드립니다!! 🙏

ohtaeg replied:
1 ~ 45개의 로또 번호를 이미 생성해서 가지고 있는 상태에서 만약 없는 번호가 들어오는 경우라면
유효하지 않은 번호라고 생각이 되었습니다 :)

그래서 validateNumber(lottoNumber)를 수행해줘도 괜찮지 않을까 하였습니다 😄

tlaqk229 replied:
부연설명 감사합니다! 이제 이해됐습니다! 말씀해주신대로 객체를 새로 생성해서 반환하는 것 코드는 의미가 없는 듯해서 validationNumber를 호출하도록 하였고 기존의 validationNumber는 판별해서 결과를 반환하는 메소드였는데, 반환값 없이 유효하지 않으면 예외처리를 하도록 수정하였고 이에 따라 수정되어야하는 부분들도 수정해보았습니다!

ohtaeg commented on src/main/java/lotto/model/domain/PurchaseCount.java:
```diff
@@ -0,0 +1,91 @@
+import java.util.Objects;
+import lotto.model.constants.ErrorMessage;
+
+public class PurchaseCount {
```
구매 갯수에 대한 vo를 만들어주셨군요 👍

ohtaeg commented on src/test/java/lotto/model/domain/LottoTest.java:
```diff
@@ -58,28 +58,28 @@ public class LottoTest {
-        lotto.addLottoNumber(new LottoNumber(4));
-        lotto.addLottoNumber(new LottoNumber(5));
-        lotto.addLottoNumber(new LottoNumber(6));
+        lotto.addLottoNumber(LottoNumber.getLottoNumberByInt(1));
+        lotto.addLottoNumber(LottoNumber.getLottoNumberByInt(2));
+        lotto.addLottoNumber(LottoNumber.getLottoNumberByInt(3));
+        lotto.addLottoNumber(LottoNumber.getLottoNumberByInt(4));
+        lotto.addLottoNumber(LottoNumber.getLottoNumberByInt(5));
+        lotto.addLottoNumber(LottoNumber.getLottoNumberByInt(6));
```
각 테스트마다 Lotto 생성을 해주고 있는데요!
이걸 메서드로 분리해주면 좋을 것 같아요 :)

각 테스트마다 다루는 로또 번호가 다르기에 행위는 같은데 다른 숫자를 다루기 위해서는
파라미터로 받아줘도 좋을 것 같네요 :)

tlaqk229 replied:
4단계 미션 피드백 영상 중에 테스트를 위한 생성자를 만드는 것을 권장한다는 자바지기님 얘기가 떠올라서 생성자를 하나 추가해서 개선해보았습니다!

ohtaeg commented on src/test/java/lotto/model/domain/LottoTest.java:
```diff
@@ -12,43 +12,43 @@ public class LottoTest {
     @Test
```
사소하지만 StringCalculatorTest 테스트가 실패하고 있는 것 같아요!
확인한번 해주시면 감사하겠습니다 😄

tlaqk229 replied:
수정했습니다!

ohtaeg reviewed (changes requested):
마지막 미션도 잘해주셨어요 👍 👍 
마지막으로 피드백을 드렸는데 확인해주시면 감사하겠습니다!!
마지막 피드백이 될 것 같아요 ㅎㅎ 화이팅입니다 🔥

tlaqk229 committed:
fix(StringCalculator): splitInputText 오류 수정

- 커스텀 구분자 선언부를 제거한 입력값을 리턴해주지 않아 오류 발생

tlaqk229 committed:
refactor(Lotto): 생성자 추가하여 테스트코드 개선

- Lotto 객체 생성, 번호추가 코드가 불필요하게 테스트코드에서
  반복적으로 작성되고 있어 생성자 추가하여 개선

tlaqk229 committed:
refactor(Lotto): 자동로또 생성 구현 위치 변경

- LottoServiceImpl에 구현되어있는 자동로또 생성 메소드를
  Lotto 객체 내부로 옮김 (도메인 객체 로직이라고 생각되어 이동)
- 불필요한 기본 생성자 제거

tlaqk229 commented:
> > 아래와 같이 생성자를 하나 더 만들고 인자로 Write 객체를 받게 하는 것과
> > 정적 팩토리 메소드에서 Write 객체를 받아 처리하는 것 사이에 차이가 있는지 의문의 들어 질문드립니다!
> > 새로운 생성자를 만들지 않고 정적 팩토리 메소드를 만들어 사용하는 경우에 얻을 수 있는 장점은 어떤 것이 있는지 궁금합니다.
> 
> 개인적으로 객체 생성시 데이터를 어디선가 getter 등을 통해 가져와서 해줘야하기 때문에 캡슐화는 큰 차이가 없다고 생각합니다 :)
> 
> 개인적으로 정적 팩토리 메서드는 생성자가 여러개가 있을 경우 효과를 발휘하는 것 같아요 :) 현업에서 일을 하다보면 많은 필드가 담긴 클래스를 다루게 되는데 필드가 많다보니 각 생성자마다 필요한 데이터들이 다르기에 여러 생성자가 생기게 됩니다. 어떤 상황에 따라 어떤 값을 가져야하는지 정적 팩터리 메서드를 통해 의사전달을 할 수 있다고 생각하며
> 
> 정적 팩토리 메서드를 사용하면 얻는 또다른 장점은 객체 생성을 호출 할 때마다 인스턴스를 새로 생성하지 않아도 되는 점이라고 생각합니다! 위에서 말씀드린 것 처럼 클래스 규모가 커서 생성 비용이 큰 객체의 경우 요청될 때마다 생성하는 것보다 정적 팩토리 메소드를 사용하면 매번 인스턴스를 새로 생성하지 않고 기존에 만들어두거나 생성한 인스턴스를 캐싱해서 재활용할 수 있는 장점이 있다고 생각합니다 :) 디자인 패턴의 플라이웨이트 패턴과 유사한 장점이 있지 않을까 합니다!
> 
> 원하시는 대답이 되었는지 모르겠네요 🙏

실제로 사용할 때 장점으로 작용하는 점들을 정리해주셔서 정적 팩토리 메소드의 사용목적이 더 잘 와닿았습니다! 도움 많이되었습니다! 감사합니다!!

ohtaeg commented on src/main/java/utils/StringCalculator.java:
```diff
@@ -33,7 +33,7 @@ private static boolean checkInputEmpty(String inputText) {
     private static String[] splitInputText(String inputText) {
         List<String> delimiters = new ArrayList<>(Arrays.asList(BASIC_DELIMITER));
-        customDelimiterHandle(inputText, delimiters);
+        inputText = customDelimiterHandle(inputText, delimiters);
```
👍

ohtaeg commented on src/test/java/lotto/model/domain/LottoTest.java:
```diff
@@ -88,23 +57,10 @@ public class LottoTest {
-        lotto2.addLottoNumber(LottoNumber.getLottoNumberByInt(9));
-        lotto2.addLottoNumber(LottoNumber.getLottoNumberByInt(10));
-
+        Lotto lotto = new Lotto(Arrays.asList(1,2,3,4,5,6));
```
👍 👍

ohtaeg reviewed (approved):
수많은 피드백을 반영하시느라 고생 많으셨어요 👍 
제가 드렸던 피드백이 윤지님께 조금이나마 도움이 되었으면 좋겠네요 :)
주셨던 질문마다 각각 코멘트를 드렸는데 마지막으로 확인만 해주시면 감사하겠습니다!
남은 미션도 화이팅하세요 🔥 고생 많으셨습니다 🙇

tlaqk229 commented:
질문과 조언 부탁드린 부분들에 대해 친절하게 답해주셔서 감사합니다! 도움이 많이 되었고, 이미 병합해주셨지만 아쉬움이 남아 피드백 주신 부분 반영해서 새로 step5_2 브랜치에 push해보았는데요. 한번더 pull request 드려도 될까요? 그동안 리뷰해주시느라 고생 많으셨습니다! 🙇  미처 생각하지 못한 부분들 많이 짚어주셔서 다양하게 생각해볼 수 있었고 성장할 수 있었던 것 같습니다! 감사합니다!!!

ohtaeg commented:
안녕하세요 윤지님!
도움이 되었다니 다행입니다!! 😊 
리뷰를 원하시면 당연히 해드리는게 인지상정입니다  👍 
이미 미션이 종료되어 슬랙에 알림이 안올거라
PR을 올려주신 후 저에게 따로 DM으로 PR링크를 공유해주시면 리뷰드리도록 할께요!!

tlaqk229 commented:
다른 분들 리뷰도 해주셔야하고 바쁘실텐데 요청 받아주셔서 감사합니다!! 🙏
